cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -600,3 +622,99 @@ func isNotFound(err error) bool {
+	// TODO: short-circuit
+
+	start := time.Now()
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return nil
+	}
+	nodeName := strings.TrimPrefix(x509cr.Subject.CommonName, "system:node:")
+	if len(nodeName) == 0 {
+		return nil
+	}
+
+	node, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// if there is no existing Node object, return success
+		return nil
+	}
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error getting node %s: %v", nodeName, err)
+	}
+
+	delete, err := shouldDeleteNode(opts, node)
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error determining if node %s should be deleted: %v", nodeName, err)
+	}
+	if !delete {
+		// if the existing Node object does not need to be removed, return success
+		return nil
+	}
+
+	err = client.CoreV1().Nodes().Delete(nodeName, &metav1.DeleteOptions{Preconditions: metav1.NewUIDPreconditions(string(node.UID))})
+	if apierrors.IsNotFound(err) {
+		// If we wanted to delete and the node is gone, this counts as success
+		return nil
+	}
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error deleting node %s: %v", nodeName, err)
+	}
+	csrApprovalStatus.WithLabelValues("node_deleted", metricsLabel).Inc()
+	csrApprovalLatency.WithLabelValues("node_deleted", metricsLabel).Observe(time.Since(start).Seconds())
+	return nil
+}
+
+func shouldDeleteNode(opts GCPConfig, node *v1.Node) (bool, error) {
+	// Newly created node might not have pod CIDR allocated yet.
+	if node.Spec.PodCIDR == "" {
+		glog.V(2).Infof("Node %q has empty podCIDR.", node.Name)
+		return false, nil
+	}
+	inst, err := getInstanceByName(opts, node.Name)
+	if err != nil {
+		if err == instanceNotFound {
+			glog.Warningf("Didn't find corresponding instance for node %q, will trigger node deletion.", node.Name)
+			return true, nil
+		}
+		glog.Errorf("Error retrieving instance %q: %v", node.Name, err)
+		return false, err
+	}
+	for _, networkInterface := range inst.NetworkInterfaces {
+		for _, r := range networkInterface.AliasIpRanges {
+			if node.Spec.PodCIDR == r.IpCidrRange {
+				glog.V(2).Infof("Instance %q has alias range that matches node's podCIDR.", inst.Name)
+				return false, nil
+			}
+			glog.Warningf("Instance %q has alias range %s that is different than node's podCIDR %s, will trigger node deletion.", inst.Name, r.IpCidrRange, node.Spec.PodCIDR)
+			return true, nil
```

Comment:
Just to confirm - there's no GKE configuration that would create Nodes with multiple AliasIPRanges on its interfaces?

The case i'm worried about is: Node has two interfaces with AliasIPRanges (or one interface with multiple ranges) and one of those ranges actually matches PodCIDR, but not the first one. In this case Node would be deleted with this logic.

Reply:
it seems like we should tolerate the PodCIDR matching any alias on the instance to be compatible with all possible usages

Reply:
I don't know of any possible usage of having multiple AliasIPRanges --- as long as we don't support multiple podCIDRs. Will bug @bowei to confirm.

Reply:
just because we don't support it today doesn't mean we never will... if any alias assigned to the instance matches, I think we should tolerate it

Reply:
Sure --- I also notice user would be able to add more alias ranges after VM creation. I will update to tolerate it.

Reply:
Added a new commit to tolerate PodCIDR matching any alias.